pandas/_libs/lib.pyx

```diff
@@ -2875,53 +2877,39 @@ def map_infer_mask(
     -------
     np.ndarray or an ExtensionArray
     """
-    cdef Py_ssize_t n = len(arr)
-    result = np.empty(n, dtype=dtype)
-
-    _map_infer_mask(
-        result,
-        arr,
-        f,
-        mask,
-        na_value,
-    )
-    if convert:
-        return maybe_convert_objects(result)
-    else:
-        return result
-
-
-@cython.boundscheck(False)
-@cython.wraparound(False)
-def _map_infer_mask(
-        ndarray[uint8_int64_object_t] out,
-        ndarray[object] arr,
-        object f,
-        const uint8_t[:] mask,
-        object na_value=no_default,
-) -> None:
-    """
-    Helper for map_infer_mask, split off to use fused types based on the result.
-    """
     cdef:
-        Py_ssize_t i, n
+        Py_ssize_t i
+        Py_ssize_t n = len(arr)
         object val
 
-    n = len(arr)
+        ndarray result = np.empty(n, dtype=dtype)
+
+        flatiter arr_it = PyArray_IterNew(arr)
+        flatiter result_it = PyArray_IterNew(result)
+
     for i in range(n):
         if mask[i]:
             if na_value is no_default:
-                val = arr[i]
+                val = PyArray_GETITEM(arr, PyArray_ITER_DATA(arr_it))
```

Comment:
What is Cython generating in this case? I'm surprised going in to the C API would make a difference

Reply:
I think this should at least avoid a speed regression taking away the ``ndarray[object]`` annotation.

I assume Cython just does manual Python-style iteration over the elements without the ndarray annotation.

Reply:
I think with the ndarray annotation it would still be using the NumPy iterators. I know you mentioned you aren't seeing a performance difference - I would just revert this unless there is something we see different in the generated code or in the benchmarks

Reply:
The no perf regression is for the new code with numpy C API iteration compared to the ``ndarray[object]`` iteration.

The benchmarks are noisy on my machine, but I'm seeing roughly a 20% slowdown just using regular indexing compared to numpy C API iteration.

Reply:
Is there an upstream issue in Cython for this?

Reply:
https://github.com/cython/cython/issues/5719 is closely related, and this approach is mentioned there.

I don't know if Cython can do much about this, though. I don't think it's easy to identify when this optimization can be made (esp. for multi-dimensional arrays), but I'm not too much of a compilers person. 

Cython is also encouraging people to use memoryviews anyways.